mvsnet/loss.py

```diff
@@ -12,48 +12,151 @@
 FLAGS = tf.app.flags.FLAGS
 
 
-def masked_loss(y_true, y_pred, interval, alpha, beta):
+def original_loss(y_true, y_pred, interval):
+    """ non zero mean absolute loss for one batch
+        This is the original loss function used in mvsnet paper"""
+    with tf.name_scope('MAE'):
+        shape = tf.shape(y_pred)
+        interval = tf.reshape(interval, [shape[0]])
+        mask_true = tf.cast(tf.not_equal(y_true, 0.0), dtype='float32')
+        denom = tf.abs(tf.reduce_sum(mask_true, axis=[1, 2, 3])) + 1e-6
+        masked_abs_error = tf.abs(
+            mask_true * (y_true - y_pred))            # 4D
+        masked_mae = tf.reduce_sum(
+            masked_abs_error, axis=[1, 2, 3])        # 1D
+        masked_mae = tf.reduce_sum((masked_mae / interval) / denom)         # 1
+    return masked_mae, tf.no_op()
+
+
+def power_loss(y_true, y_pred, interval, alpha, beta, no_interval_norm=False):
     """ non zero mean absolute loss for one batch
 
     This function parameterizes a loss of the general form:
 
     Loss = N * (|y_true-y_pred| + epsilon(y_true))^alpha / y_true^beta
 
-    where alpha and beta are scalars, and N is a normalization constant which depends on 
+    where alpha and beta are scalars, and N is a normalization constant which depends on
     alpha, beta and y_true. epsilon(y_true) is the expected noise of the measurement of y_true, and helps to prevent overfitting to noise
-    in the depth map. 
+    in the depth map. This is the loss for an individual pixel. The total loss for a depth map
+    is the average taken over all valid pixels.
     Additionally the numerator and denominator are multipled by a mask to mask out
     invalid pixels in the label. This was omitted above for notational simplicity.
 
     See this paper for a description and analysis of the noise model of the kinect sensor
     -- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3304120/
 
-    One key takeaway is that the random error in Kinect depth maps increases quadratically with distance and 
+    One key takeaway is that the random error in Kinect depth maps increases quadratically with distance and
     reaches a maximum of 4cm at the maximum range of 5 meters
      """
 
     with tf.name_scope('MAE'):
         shape = tf.shape(y_pred)
         interval = tf.reshape(interval, [shape[0]])
+        # mask_true is a tensor of 0's and 1's, where 1 is for valid pixels and 0 for invalid pixels
         mask_true = tf.cast(tf.not_equal(y_true, 0.0), dtype='float32')
-        denominator = tf.abs(tf.reduce_sum(mask_true, axis=[1, 2, 3])) + 1e-6
-        if beta != 1.0:
-            denominator = tf.math.pow(denominator, beta)
-        # Below we assume the random error in y_true increases linearly with distance
-        # and that the error in y_true is 5mm at a distance of 1 meter
+        # The number of valid pixels in the depth map -- used for taking the average over only valid pixels
```

Comment:
not sure about the specific dataset you're using, but is there a value that's set when the depth value is too high, e.g. for the sky? do we want to mask out those pixels as well?

Reply:
invalid depths (could be because its too high or whatever), are all set to a value of 0